chain-crypto/src/digest.rs

```diff
@@ -242,3 +246,80 @@ impl<H: DigestAlg> fmt::Debug for Digest<H> {
         f.write_str(")")
     }
 }
+
+impl<H: DigestAlg> Digest<H> {
+    /// Get the digest of a slice of data
+    pub fn digest(slice: &[u8]) -> Self {
+        let mut ctx = Context::new();
+        ctx.append_data(slice);
+        ctx.finalize()
+    }
+}
+
+use std::marker::PhantomData;
+
+/// A typed version of Digest
+pub struct DigestOf<H: DigestAlg, T> {
+    inner: Digest<H>,
+    marker: PhantomData<T>,
+}
+
+impl<H: DigestAlg, T> Clone for DigestOf<H, T> {
+    fn clone(&self) -> Self {
+        DigestOf {
+            inner: self.inner.clone(),
+            marker: self.marker.clone(),
+        }
+    }
+}
+
+impl<H: DigestAlg, T> From<DigestOf<H, T>> for Digest<H> {
+    fn from(d: DigestOf<H, T>) -> Self {
+        d.inner
+    }
+}
+
+impl<H: DigestAlg, T> From<Digest<H>> for DigestOf<H, T> {
+    fn from(d: Digest<H>) -> Self {
+        DigestOf {
+            inner: d,
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<H: DigestAlg, T> DigestOf<H, T> {
+    /// Coerce a digest of T, to a digest of U
+    pub fn coerce<U>(&self) -> DigestOf<H, U> {
+        DigestOf {
+            inner: self.inner.clone(),
+            marker: PhantomData,
+        }
+    }
+
+    /// Get the digest of object T, given its AsRef<[u8]> implementation
+    pub fn digest(obj: &T) -> Self
+    where
+        T: AsRef<[u8]>,
+    {
+        let mut ctx = Context::new();
+        ctx.append_data(obj.as_ref());
+        DigestOf {
+            inner: ctx.finalize(),
+            marker: PhantomData,
+        }
+    }
+
+    /// Get the digest of object T, given its serialization function in closure
+    pub fn digest_with<F>(obj: &T, f: F) -> Self
+    where
+        F: FnOnce(&T) -> &[u8],
```

Comment:
I think this is problematic for several reasons.

The closure's signature assumes there is a contiguous byte slice inside the object and that is the only content that needs to be hashed. There really should be a trait for that and the types define the way they are hashed with.

Also, with this function there is no invariant to how `DigestOf` has been calculated, potentially leading to bugs down the line where digests are compared.

I once designed [a little crate](https://crates.io/crates/digest-hash) to deal with cryptographically stable hashing; check out the design.

Reply:
yes, it's far from perfect to assume the presence of a slice of byte, this can be improved later; At the same time, it's really a type poorness to assume that any object can be hashed here, and not only sliced of data (think `MessageRaw`, `HeaderRaw`)

Reply:
There is no technical issue here that can't be solved with a `Hash`-like trait on `T`. Unless you really want different custom ways of hashing the same type.